Document the Feasibility Process

alexanderkiel opened the pull request:
Closes: #4

alexanderkiel committed:
Document the Feasibility Process

Closes: #4

alexanderkiel committed:
Document the Feasibility Process

Closes: #4

DiCanio commented on codex-process-feasibility/README.md:
```diff
@@ -0,0 +1,58 @@
+
+After the ZARS FHIR server receives the resources, it notifies the ZARS BPE via websocket subscription by transmitting the Task resource (2). The incoming Task resource starts the `request` process that implements the query distribution to all appropriate DIZ'es. To each DIZ, an `execute` message (3) is sent via a Task resource.
+
+After arriving at the DIZ FHIR communication server, the `execure` Task resource is transferred to the DIZ BPE via  websocket subscription (4), starting the `execute` prozess.
```
```suggestion
After arriving at the DIZ FHIR communication server, the `execute` Task resource is transferred to the DIZ BPE via  websocket subscription (4), starting the `execute` process.
```

DiCanio commented on codex-process-feasibility/README.md:
```diff
@@ -0,0 +1,58 @@
+
+![fig-1](./docs/requestSimpleFeasibility.png)
+
+In the BPMN model of the `request` process, the start message it the `request` message. After selecting the request targets (the DIZ'es), the `execute` messages are send in the next step. After that a subprocess is started for each target, which will wait for the `result` messages to arrive.
```
What does BPMN stand for? Acronym usage without introduction.

alexanderkiel replied:
Added Full Name and Wikipedia Link

DiCanio commented on codex-process-feasibility/README.md:
```diff
@@ -0,0 +1,58 @@
+
+![fig-1](./docs/requestSimpleFeasibility.png)
+
+In the BPMN model of the `request` process, the start message it the `request` message. After selecting the request targets (the DIZ'es), the `execute` messages are send in the next step. After that a subprocess is started for each target, which will wait for the `result` messages to arrive.
```
```suggestion
In the BPMN model of the `request` process, the start message is the `request` message. After selecting the request targets (the DIZ'es), the `execute` messages are send in the next step. After that a subprocess is started for each target, which will wait for the `result` messages to arrive.
```

DiCanio commented on codex-process-feasibility/README.md:
```diff
@@ -0,0 +1,58 @@
+## Request and Execute Process in Detail
+
+Messages, queries and results are represented by FHIR resources. The following three figures show the entire FHIR resource flow. In difference to the figure above, only one DIZ is represented, but both, the ZARS and the DIZ are divided into its individual components. On the ZARS side the components are the FHIR communication server and the BPE, were the DIZ contains an additional Blaze FHIR serve ...
+
```
Maybe add another subheading like `Distribute Query`.

DiCanio commented on codex-process-feasibility/README.md:
```diff
@@ -0,0 +1,58 @@
+After the ZARS FHIR server receives the resources, it notifies the ZARS BPE via websocket subscription by transmitting the Task resource (2). The incoming Task resource starts the `request` process that implements the query distribution to all appropriate DIZ'es. To each DIZ, an `execute` message (3) is sent via a Task resource.
+
+After arriving at the DIZ FHIR communication server, the `execure` Task resource is transferred to the DIZ BPE via  websocket subscription (4), starting the `execute` prozess.
+
```
Maybe add another subheading like `Execute Query`.

DiCanio commented on codex-process-feasibility/README.md:
```diff
@@ -0,0 +1,58 @@
+After receiving the MeasureReport, the DIZ BPE stores it on the DIZ FHIR communication server (8) in order to make it available for the ZARS.
+
+In its last step, the `execute` process sends a `result` message (9) to the ZARS. The `result` message references the MeasureReport, so that it can be retrieved by the ZARS.
+
```
Maybe add another subheading like `Retrieve Results`.

DiCanio commented on codex-process-feasibility/README.md:
```diff
@@ -0,0 +1,58 @@
+# NUM CODEX Feasibility Process
+
+The feasibility process is used to run a distributed feasibility query. A feasibility query contains multiple criteria and returns one population count per organization. The feasibility query is requested by one organization (usually the Zentrale Antrags- und Registerstelle (ZARS)) and executed in multiple organizations (usually Datenintegrationszentren (DIZ)).
```
Introduce acronyms using a boldface to make them stand out.

Something like:
[…] usually the **Zentrale Antrags- und Registerstelle (ZARS)** […]

alexanderkiel replied:
I don't think so.

DiCanio commented on codex-process-feasibility/README.md:
```diff
@@ -0,0 +1,58 @@
+# NUM CODEX Feasibility Process
+
+The feasibility process is used to run a distributed feasibility query. A feasibility query contains multiple criteria and returns one population count per organization. The feasibility query is requested by one organization (usually the Zentrale Antrags- und Registerstelle (ZARS)) and executed in multiple organizations (usually Datenintegrationszentren (DIZ)).
```
Always try to go for active voice instead of passive voice.

**Why?:**
- Active voice tends to be shorter and more precise.
- Passive voice obfuscates ideas and turns sentences on their head (mental load).
- Readers typically translate passive to active voice in their head.

**How?:**
Avoid sentence structures in the form of `form of be + past participle verb`.

**Example:**
```suggestion
A distributed feasibility query runs based on the feasibility process. A feasibility query contains multiple criteria and returns one population count per organization. An organization (usually the **Zentrale Antrags- und Registerstelle (ZARS)**) requests a feasibility query to start off the process. Once started multiple organizations (usually **Datenintegrationszentren (DIZ)**) execute this query and report results back.
```

DiCanio commented on codex-process-feasibility/README.md:
```diff
@@ -0,0 +1,58 @@
+
+![fig-1](./docs/codex-feasibility-process-01.png)
+
+As you can see in the figure above, the feasibility query process starts with a `request` message (1) that the ZARS or any other initiating organization sends to itself. By sending this message to the ZARS itself, the business process engine (BPE) of the ZARS can be used to distribute the `execute` messages (2) to all DIZ'es in question. After receiving the `execute` messages, each DIZ will calcu ...
```
Do not describe processes in prose! Use numbered lists instead.
When using numbered lists make sure the content is `parallel`, i.e. **all** list entries are either:
- single words
- full sentences
- short phrases

The important thing is that the reader can recognize them as coherent.

The use of numbered lists is also beneficial for jumping quickly between illustration and explanation. Inline references like (3) are hard to find.

DiCanio replied:
Applies to all the other descriptions as well. Has only been vicariously picked.

DiCanio commented on codex-process-feasibility/README.md:
```diff
@@ -0,0 +1,58 @@
+
+![fig-1](./docs/executeSimpleFeasibility.png)
+
+The BPMN model of the `execute` process is strait forward an already explained in detail above.
```
```suggestion
The BPMN model of the `execute` process is straightforward and already explained in detail above.
```

DiCanio reviewed (commented):
The description of the whole process is very clear and helpful 👍🏻 

I only have a few concerns about the document's readability and the kind of technical writing in general. See review comments for suggestions.

alexanderkiel committed:
Fix Typos

Co-authored-by: DiCanio <DiCanio@users.noreply.github.com>

alexanderkiel committed:
Added BPMN Full Name and Wikipedia Link

alexanderkiel committed:
Fix Typo

alexanderkiel committed:
Add Subsections

alexanderkiel committed:
Use Active Voice

alexanderkiel committed:
Use Numbered Lists instead of Prose

Also updated step 12 to use a transaction bundle to update Task and MeasureReport in one transaction.

alexanderkiel committed:
Fix Typo

Co-authored-by: DiCanio <DiCanio@users.noreply.github.com>

AndreaToth reviewed (approved):
Very instructive and valuable for understanding the process. It would be nice to have an example FHIR CQL query here with result.
